packages/core/src/internal/fetchSource.ts

```diff
@@ -1,68 +1,165 @@
+
+const asyncIterator =
+  typeof Symbol !== 'undefined' ? Symbol.asyncIterator : null;
+const decoder = typeof TextDecoder !== 'undefined' ? new TextDecoder() : null;
+const jsonHeaderRe = /content-type:[^\r\n]*application\/json/i;
+const boundaryHeaderRe = /boundary="?([^=";]+)"?/i;
+
+type ChunkData = { done: false; value: Buffer | Uint8Array } | { done: true };
 
-const executeFetch = (
+interface AbstractReader {
+  next?(): Promise<ChunkData>;
+  read?(): Promise<ChunkData>;
+  cancel?(): void;
+}
+
+const toString = (input: Buffer | ArrayBuffer): string =>
+  typeof Buffer !== 'undefined' && Buffer.isBuffer(input)
+    ? input.toString()
+    : decoder!.decode(input);
+
+const executeIncrementalFetch = (
+  onResult: (result: OperationResult) => void,
   operation: Operation,
-  url: string,
-  fetchOptions: RequestInit
-): Promise<OperationResult> => {
-  const fetcher = operation.context.fetch;
+  response: Response
+): Promise<void> => {
+  const contentType =
+    (response.headers &&
+      (response.headers.get('Content-Type') ||
+        response.headers['Content-Type'])) ||
+    '';
+  if (!/multipart\/mixed/i.test(contentType)) {
+    return response.json().then(payload => {
+      onResult(makeResult(operation, payload, response));
+    });
+  }
+
+  let boundary = '-';
+  const boundaryHeader = contentType.match(boundaryHeaderRe);
+  if (boundaryHeader) boundary = boundaryHeader[1];
+
+  let iterator: AbstractReader;
+  if (asyncIterator && response[asyncIterator]) {
+    iterator = response[asyncIterator]();
+  } else if ('body' in response && response.body) {
+    iterator = response.body.getReader();
+  } else {
+    throw new TypeError('Streaming requests unsupported');
+  }
 
-  let statusNotOk = false;
-  let response: Response;
-
-  return (fetcher || fetch)(url, fetchOptions)
-    .then((res: Response) => {
-      response = res;
-      statusNotOk =
-        res.status < 200 ||
-        res.status >= (fetchOptions.redirect === 'manual' ? 400 : 300);
-      return res.json();
-    })
-    .then((result: any) => {
-      if (!('data' in result) && !('errors' in result)) {
-        throw new Error('No Content');
+  let buffer = '';
+  let isPreamble = true;
+  let nextResult: OperationResult | null = null;
+  let prevResult: OperationResult | null = null;
+
+  function next(data: ChunkData): Promise<void> | void {
+    if (!data.done) {
+      const chunk = toString(data.value);
+
+      let boundaryIndex = chunk.indexOf(boundary);
+      if (boundaryIndex > -1) {
+        boundaryIndex += buffer.length;
+      } else {
+        boundaryIndex = buffer.indexOf(boundary);
       }
 
-      return makeResult(operation, result, response);
-    })
-    .catch((error: Error) => {
-      if (error.name !== 'AbortError') {
-        return makeErrorResult(
-          operation,
-          statusNotOk ? new Error(response.statusText) : error,
-          response
-        );
+      buffer += chunk;
```

Comment:
can try mention this code was largely copied from my work (https://github.com/maraisr/meros/blob/d914f7461e2f3d0e3c2064c91ac95c28ef1e79d4/src/lib/browser.ts#L34)

Reply:
yea, of course, we're just internally discussing what an attribution comment should look like

Reply:
no need — this is OSS. You don't HAVE to do anything 😅 just a little upset my lib wasnt good enough to have just been used. Ill be watching this very closely to improve my things <3

Reply:
Oh, either way, attribution fully necessary as you drive the RFC. It'll also make it easier for external contributors to spot divergences for us and compare them. I'd love to keep it aligned going forward, but this is the easiest approach, so far ✨

One part of the reason to diverge is to allow isomorphic use-cases. I've seen the `package.json` switch and generally have made some bad experiences with those. That's why the persisted fetch exchange doesn't use it for Node/Browser and instead uses an eval-require. 😅 I've realised that there are so many commonalities in both files that we can attempt to skip the differences.

The other part is IE11 support. We'd like to be able to "partially support" it at least. Meaning that we won't fiddle around with XMLHttpRequest (especially since that deprecates the configurable `fetch` and is hence even more confusing) but because we'd have to transpile the generator and `for await ... of` code, which wouldn't yield compact code.

Speaking of compact, we're just generally aiming to shave off bytes and keep things simple. This approach doesn't change our stack a lot and we have some control to golf some bytes and reduce the bundle size impact. While this point is admittedly not that important in general, we do try to keep things small, especially when adding new features

Reply:
Think the idea was — and still in flux. Node is kept in buffer as opposed to operating on strings. Re the async iterables — would it makes sense for me to expose a potentially `meros/legacy` or something to that effect, where its implemented as a callback? For me its more about make sure that my other 3.5M downloaded users benefit from patches and bug fixes that may result from findings discovered here. 

Also re the isomorphic nature, honestly can say this isnt an issue. unpkg correctly handles it, esbuild and webpack also. But for the unknown browserify or whatever bundler, i could forgo performance here and expose a `meros/isomorphic` sub-module that merges the 2 behaviors.

Ultimately feel we should still try and work together here, rather than go down this forked path. 

I'm open for amends/changes/bends/alignment to suffice urql's needs.

Reply:
I mean, don't get me wrong, I think `meros/legacy` would be neat and there are compromises to be had, but even all things considered currently we keep the dependencies we take on at practically zero either way; the big exception was `@graphql-typed-document-node/core` since TypeScript needs this to mostly be a deduplicated, unique package. So if we hypothetically said that everything was as we needed it in `meros` (which for the most part, they are of course) we'd still be reluctant to take on an extra dependency.

That's not a direct attack at other maintainers, it's mostly for practical reasons. For the one part, it does allow us to have full control over our build output and that allows us to code golf and uniformly keep sizes small and the experience coherent. One big point we picked up from the ecosystem, especially in the early days of GraphQL, is that any given library picks up tens of dependencies to achieve one thing. And while that doesn't add up to any difference in the quality of what we're building, it's important to us to outwardly show what we stand for with stats.

Apart from that however, we believe in alignment through specifications, throughout the GraphQL ecosystem. We're big fans of the GraphQL over HTTP spec-extension in general, and weren't happy how we had to approach `@urql/exchange-persisted-fetch` and `@urql/exchange-multipart-fetch` with a vague, coincidental convention.
This then obviously leaves us with questions of what happens when a spec does emerge and we're faced with either maintaining an implementation or integrating an external library.

With `graphql-ws`, this choice was rather easy. Since the beginning, there wasn't a spec to align but there was a de-facto practical conventional package in `subscriptions-transport-ws`. We decided to align loosely on that, later seeing that other vendors like AppSync were following even different practices, which then worked out. Now that a spec is emerging this will likely not change, for the simple reason that most people have to opt into subscriptions either way, and this isn't a super common choice we see.

With this decision it's a little different. It's impossible for us to _not_ integrate defer/stream closely as it influences our core implementation. On the other hand, as you could see, we also didn't want to issue a breaking change that replaces our pluggable `fetch` option, so we're stuck with making some assumptions here. This includes that we may see some breakage from users that are using `unfetch` and are trying this out, which means we'd like to control in the future how we surface this (similar to the risk of IE11), and only as a last resort consider deprecating the `fetch` option.

It's also possible that we'd like to diverge a little from `meros` in general (outside of spec concerns) every now and then. By having the code ourselves, not only do we get benefits in our output and little to no overhead in integrating a library and duplicating some code, when something goes wrong (say, with React Native, or a new bundler) we can act fast and still contribute changes back.

So, it's not a matter of not contributing back or not working together, which we still will and want to do, it's just that we don't see collaboration as effective in shared code at all times, and compromising on a couple of things that we see as necessary. In fact, I believe we'd likely work together more if we're not stuck putting all our eggs in one basket, which leads to a situation where all roads lead to ~rome~ one library, but rather to shared issues and spec changes 😄